crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,406 @@
+use std::{alloc::Layout, cell::UnsafeCell, marker::PhantomData, ptr::NonNull};
+
+#[derive(Debug)]
+pub struct BlobVec {
+    item_layout: Layout,
+    capacity: usize,
+    len: usize,
+    data: UnsafeCell<NonNull<u8>>,
+    swap_scratch: UnsafeCell<NonNull<u8>>,
+    drop: unsafe fn(*mut u8),
+}
+
+impl BlobVec {
+    pub fn new(item_layout: Layout, drop: unsafe fn(*mut u8), capacity: usize) -> BlobVec {
+        unsafe {
+            let mut blob_vec = BlobVec {
+                swap_scratch: UnsafeCell::new(NonNull::dangling()),
+                item_layout,
+                capacity: 0,
+                len: 0,
+                data: UnsafeCell::new(NonNull::dangling()),
+                drop,
+            };
+            if item_layout.size() == 0 {
+                blob_vec.capacity = usize::MAX;
+            } else {
+                blob_vec.swap_scratch =
+                    UnsafeCell::new(NonNull::new(std::alloc::alloc(item_layout)).unwrap());
+                blob_vec.reserve(capacity);
+            }
+            blob_vec
+        }
+    }
+
+    pub fn new_typed<T>(capacity: usize) -> BlobVec {
+        BlobVec::new(Layout::new::<T>(), drop_ptr::<T>, capacity)
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    pub fn reserve(&mut self, amount: usize) {
+        let available_space = self.capacity - self.len;
+        if available_space < amount {
+            self.grow(amount - available_space);
+        }
+    }
+
+    fn grow(&mut self, increment: usize) {
+        debug_assert!(self.item_layout.size() != 0);
+
+        let new_capacity = self.capacity + increment;
+        unsafe {
+            let new_data = if self.capacity == 0 {
+                std::alloc::alloc(Layout::from_size_align_unchecked(
+                    self.item_layout.size() * new_capacity,
+                    self.item_layout.align(),
+                ))
+            } else {
+                std::alloc::realloc(
+                    (*self.data.get()).as_ptr(),
+                    Layout::from_size_align_unchecked(
+                        self.item_layout.size() * self.capacity,
```

Comment:
The unstable `layout.reserve` uses the following impl:

```rust
// This cannot overflow. Quoting from the invariant of Layout:
// > `size`, when rounded up to the nearest multiple of `align`,
// > must not overflow (i.e., the rounded value must be less than
// > `usize::MAX`)
let padded_size = self.size() + self.padding_needed_for(self.align());
let alloc_size = padded_size.checked_mul(n).ok_or(LayoutError { private: () })?;

// SAFETY: self.align is already known to be valid and alloc_size has been
// padded already.
unsafe { Ok((Layout::from_size_align_unchecked(alloc_size, self.align()), padded_size)) }
```

The current code forgets to apply the alignment and check for overflows. Same problem in the Drop impl.

Reply:
Hmm does this mean the RawVec impl in the Nomicon is incorrect? Or do the `heap` apis do this on our behalf?
https://doc.rust-lang.org/nomicon/vec-alloc.html

I'm guessing we can just use the stable Layout::array, which is what the "real" RawVec uses internally?

Reply:
Ah wait that takes a type, which we can do here :)

Reply:
> The unstable layout.reserve uses the following impl:

Oops, s/reserve/repeat

> I'm guessing we can just use the stable Layout::array, which is what the "real" RawVec uses internally?

`Layout::array` uses `Layout::repeat` internally. `Layout::array` doesn't work here as it accepts a type parameter and not an arbitrary `Layout` at runtime. The RawVec impl in the Nomicon is correct I think as I don't think any rust types have a size < alignment. `Layout` does allow this though.